Added a plugin for fluxible to add copy methods to context

PeteJodo opened the pull request:
### A Proposal for Copy Methods on Context

A fluxible plugin that would add the additional methods: `getCopy` and `extendCopy` to the contexts: `ComponentContext`, `ActionContext`, and `StoreContext` (**note:** not on FluxibleContext).

This would allow easy access to copy for react component, actions, and stores without having to pass it down through props and payloads. It's also easy to start using, using the `message-center.js` entry file from adminv2 as an example, all that's needed is:
```javascript
// ...
messageCenterApp.plug(require('copy-text/plugins/fluxible/copyTextPlugin'));
var context = messageCenterApp.createContext({ svCopyPath: 'mc.copy' });
// ...
```

Some catches downstream, when using `provideContext`, a second parameter should be passed that looks like...
```javascript
provideContext(SomeTopLevelComponent, {
    getCopy: React.PropTypes.function.isRequired,
    extendCopy: React.PropTypes.function.isRequired
});
```

Maybe I can add to the PR a helper that exports that object.

Also a user should add `contextTypes` to their component i.e.:
```javascript
SomeComponent.contextTypes = {
    executeAction: React.PropTypes.function.isRequired,
    getStore: React.PropTypes.function.isRequired,
    getCopy: React.PropTypes.function.isRequired,
    extendCopy: React.PropTypes.function.isRequired
};
```

Users only need to add the contextTypes that their component actually uses to reduce that boilerplate.

And that's all that's required to use the plugin....
```javascript
render() {
    return <h1>{this.context.getCopy('myTitleKey')}</h1>;
}
```

##### Additional notes
- I didn't make this it's own module since it's related to copy-text and also copy-text requires `dibs-endpoints` which makes this have to be closed-source...no point using up another closed-source repo for our quota
- not sure about `extendCopy` since each context instance gets it own copyText object, extending copy within a component would not be visible to actions **BUT** would be visible to child components (I'm pretty sure, since it's the same context getting passed down). Looking for feedback on that. Maybe not even include an `extendCopy` method?
- this PR is incomplete, still need to do/write tests

+@robrichard
+@jrdrg 
+@twhid

antialias commented on plugins/fluxible/copyTextPlugin.js:
```diff
@@ -0,0 +1,28 @@
+            }
+            if (options.globalCopy) {
+                copyText.addGlobalCopy(options.globalCopy);
+            }
```
I don't think these options are necessary. If copyRext needs global configuration, that can be done on the module directly.

PeteJodo replied:
yeah that's true. Could be a little weird, taking my example from before...
```javascript
// ...
require('copy-text').addGlobalSVPath('mc.copy');
messageCenterApp.plug(require('copyText/plugins/fluxible/copyTextPlugin'));
var context = messageCenterApp.createContext();
// ...
```
Regardless, the options aren't required so a user has the option of doing either way. I can remove it though

robrichard commented:
cool idea :+1:

PeteJodo commented:
@antialias 
I removed the options part for defining and global copy and SV copy. The only issue with that is that it doesn't work for modules using the `bunsen/helpers/copyText` so when using this, we would have to make sure to remove all instances of `bunsen/helpers/copyText` and use `copy-text`.

Also I added some helpers `provideContextWithCopy` (so people don't have to define the custom context types for copy every time for the second parameter to `provideContext`) and `copyContextTypes` so when defining a component's `contextTypes` they can just do:
```javascript
SomeComponent.contextTypes = Object.assign({
    executeAction: React.PropTypes.func.isRequired
}, require('copy-text/plugins/fluxible/helpers/copyContextTypes'));
```

The problem I have with it though is that it adds `react` and `fluxible` as dependencies to copy-text which is not necessary. I have an idea which is to do something similar as what fluxible does with their [repo](https://github.com/yahoo/fluxible) which is put the `copy-text` code into it's own folder with it's own package.json and that could be the `copy-text` module then have the plugins folder be another npm module and then at the top-level  it would just be the `copy-text-repo` module.

Or maybe it isn't worth and I shouldn't even include the helpers. What do you think?

robrichard commented:
I don't like the helper for contextTypes, I think each component should define exactly what it needs (either getCopy, extendCopy or both if it uses both)

antialias commented:
> The problem I have with it though is that it adds react and fluxible as dependencies to copy-text which is not necessary.

That's solvable by making this plugin its own package that depends on copy-text, react, and fluxible.

PeteJodo commented:
@robrichard I agree with that actually; going to remove it.
@antialias That's what I suggested, I just don't want to create a new github repo so I was going to that right within this repo

antialias commented:
> I just don't want to create a new github repo

I think you should create a new repo, or at least a new package somehow for the sake of keeping the dependencies in check. We have an unlimited number of public repos, and we're going to make copy-text public soon, so don't worry about depending on a private repo.

PeteJodo commented:
Breaking this out of copy-text into it's own repo. Closing.

PeteJodo commented:
@antialias fyi it looks like copy-text is already public

antialias commented:
ok, maybe I was thinking about endpoints.
